ndk/src/midi.rs

```diff
@@ -0,0 +1,286 @@
+use std::fmt;
+use std::marker::PhantomData;
+use std::os::raw::{c_int, c_uint};
+use std::ptr::NonNull;
+
+/// Result of [`MidiOutputPort::receive`].
+#[derive(Copy, Clone, Debug)]
+#[repr(u32)]
+pub enum MidiOpcode {
+    /// No MIDI messages are available.
+    NoMessage,
+    /// Received a MIDI message with the given length.
+    Data(usize),
+    /// Instructed to discard all pending MIDI data.
+    Flush,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, TryFromPrimitive, IntoPrimitive)]
+#[repr(u32)]
+pub enum MidiDeviceType {
+    Bluetooth = ffi::AMIDI_DEVICE_TYPE_BLUETOOTH,
+    USB = ffi::AMIDI_DEVICE_TYPE_USB,
+    Virtual = ffi::AMIDI_DEVICE_TYPE_VIRTUAL,
+}
+
+#[derive(Debug)]
+pub struct MidiDevice {
+    inner: NonNull<ffi::AMidiDevice>,
+}
+
+impl MidiDevice {
+    /// Creates an `MidiDevice` from a pointer.
+    ///
+    /// # Safety
+    /// By calling this function, you assert that the pointer is a valid pointer to a native
+    /// `AMidiDevice`.
+    pub unsafe fn from_ptr(inner: NonNull<ffi::AMidiDevice>) -> Self {
+        Self { inner }
+    }
+
+    fn as_ptr(&self) -> *mut ffi::AMidiDevice {
+        self.inner.as_ptr()
+    }
+
+    /// Connects a native Midi Device object to the associated Java MidiDevice object.
+    ///
+    /// Use this AMidiDevice to access the rest of the native MIDI API.
+    pub fn from_java(
+        env: *mut jni_sys::JNIEnv,
+        midi_device_obj: jni_sys::jobject,
+    ) -> Result<MidiDevice> {
+        unsafe {
+            let ptr = construct(|res| ffi::AMidiDevice_fromJava(env, midi_device_obj, res))?;
+            Ok(Self::from_ptr(NonNull::new_unchecked(ptr)))
+        }
+    }
+
+    /// Gets the number of input (sending) ports available on the specified MIDI device.
+    pub fn num_input_ports(&self) -> Result<usize> {
+        let num_input_ports = unsafe { ffi::AMidiDevice_getNumInputPorts(self.as_ptr()) };
+        if num_input_ports >= 0 {
+            Ok(num_input_ports as usize)
+        } else {
+            NdkMediaError::from_status(ffi::media_status_t(num_input_ports as c_int)).map(|_| 0)
+        }
+    }
+
+    /// Gets the number of output (receiving) ports available on the specified MIDI device.
+    pub fn num_output_ports(&self) -> Result<usize> {
+        let num_output_ports = unsafe { ffi::AMidiDevice_getNumOutputPorts(self.as_ptr()) };
+        if num_output_ports >= 0 {
+            Ok(num_output_ports as usize)
+        } else {
+            Err(
+                NdkMediaError::from_status(ffi::media_status_t(num_output_ports as c_int))
+                    .unwrap_err(),
+            )
+        }
+    }
+
+    /// Gets the MIDI device type.
+    pub fn device_type(&self) -> Result<MidiDeviceType> {
+        let device_type = unsafe { ffi::AMidiDevice_getType(self.as_ptr()) };
+        if device_type >= 0 {
+            let device_type =
+                MidiDeviceType::try_from_primitive(device_type as u32).map_err(|e| {
```

Comment:
Other code seems to use `try_from()` instead of `try_from_primitive()`?

Reply:
Fixed!